### PR TITLE
WIP: arbitrary continuous distribution

### PR DESCRIPTION
**Pull request by andyfaff**
I'm putting this up as a WIP. I mentioned this on the dev mailing list a [while ago](https://mail.scipy.org/pipermail/scipy-dev/2016-April/021307.html).
Basically I'd like a class to define an arbitrary continuous distribution. The continuous distribution is created from one of the following:
- a function
- samples from a function
- a histogram

At the moment it's still in early stages, but I'm putting it up for comments so that I don't go right off the rails.

### CODE REVIEW

**Comment by rgommers** [comment]
It may be useful to first add some more context about what this really attempt to do to the docstring. 

When given a function, that function is simply scaled and treated as the ``pdf``. This looks useful to me. It will only work for functions with a finite integral over their support though, which is not mentioned or checked for.

When given samples/histogram, this is a fairly crude approximation. A discussion of limitations/alternatives seems necessary. Choosing an interpolating instead of approximating spline will give rise to weird shapes, shouldn't be used for noisy data. When you have samples, a better choice in most circumstances would be to get the pdf with ``gaussian_kde`` for example, and then using that as ``_pdf`` and deriving all other properties of the distribution from it.

File: scipy/stats/arb_dist.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    y : np.array or callable
+        Specifies the probability distribution function. Can be specified as an
+        array or as a function returning a single value. If `y` is an array
+        then the distribution is a sampled PDF and should have a corresponding
+        `x` value. Note that the area under the PDF is normalised to 1, i.e.
+        ``rv_arbitrary.cdf(np.max(x)) == 1``. All values in `y` should be
+        ``y >= 0``.
+    x : array_like
+        Range of support for the distribution. If `y` is a function then the
+        distribution is defined over ``[np.min(x), np.max(x)]``. If `y` is an
```
Comment (rgommers) [inline scipy/stats/arb_dist.py]: ``x`` should already be a sequence of length 2 in this case, I wouldn't allow longer input for clarity.

File: scipy/stats/arb_dist.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    interpolation : int, optional
+        If `y` represents a non-histogrammed sampled PDF (i.e.
+        ``y.size == x.size``), then `interpolation` specifies the order of an
+        interpolating spline between each of the sampled points. The default of
+        1 represents linear interpolation, 2 is quadratic interpolation, etc.
+        This value should be an integer in the range [1-5].
+
+    Examples
+    --------
+    Let us consider a quadratic distribution specified by a function:
```
Comment (rgommers) [inline scipy/stats/arb_dist.py]: For this to render correctly, there always needs to be a blank line between text and lines starting with ``>>>``

File: scipy/stats/arb_dist.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        a histogram, and the PDF will be step-like.
+    xtol : float, optional
+        The tolerance for fixed point calculation for generic ppf.
+    interpolation : int, optional
+        If `y` represents a non-histogrammed sampled PDF (i.e.
+        ``y.size == x.size``), then `interpolation` specifies the order of an
+        interpolating spline between each of the sampled points. The default of
+        1 represents linear interpolation, 2 is quadratic interpolation, etc.
+        This value should be an integer in the range [1-5].
+
```
Comment (rgommers) [inline scipy/stats/arb_dist.py]: A ``See Also`` section would be useful here. ``rv_sample`` for the discrete equivalent of this, ``gaussian_kde``, ``numpy.histogram``, ``statsmodels`` KDE/histogram.

File: scipy/stats/arb_dist.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        ``y >= 0``.
+    x : array_like
+        Range of support for the distribution. If `y` is a function then the
+        distribution is defined over ``[np.min(x), np.max(x)]``. If `y` is an
+        array then the distribution represents a sampled PDF function. If
+        ``y.size == x.size``, it is assumed the PDF is sampled at each of the
+        `x` values. If ``y.size + 1 == x.size`` then the PDF is assumed to be
+        a histogram, and the PDF will be step-like.
+    xtol : float, optional
+        The tolerance for fixed point calculation for generic ppf.
```
Comment (rgommers) [inline scipy/stats/arb_dist.py]: missing ``seed`` in parameter list

File: scipy/stats/arb_dist.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+            elif szx == szy + 1:
+                self._init_hist()
+            else:
+                raise ValueError("The x.size must equal y.size for a"
+                                 " sampled function, or y.size + 1 for a"
+                                 " sampled histogram")
+
+        super(rv_arbitrary, self).__init__(a=np.min(self._x),
+                                           b=np.max(self._x),
+                                           xtol=xtol, seed=seed)
```
Comment (rgommers) [inline scipy/stats/arb_dist.py]: This will need to avoid the messing with the docstring done at the end of ``rv_continuous.__init__``.

**Comment by ev-br** [comment]
This tries to address a very real problem any Monte Carlo work faces sooner or later (IME, sooner than later and many times over), so the need is really there. However --- I don't want to sound negative, but nevertheless --- I'm not sure this is the right way of addressing it.

IMO, having a one-size-catch-all subclass of `rv_continuous` 

* offers too little advantage over subclassing `rv_continuous` in the user code
* adds too much complexity for a user to understand how to use this class
* does not catch many use cases. 
* holds a user's hand where it shouldn't IMO --- e.g. normalizing a PDF is a user's job
* but fails to help if the PDF is a combination of a tabulation and asymptotics which is known in a functional form.
* it does not give enough flexibility for choosing an interpolator if the PDF is tabulated: as Ralf noted already, if the samples are noisy, you want a least-squares fit; if you know the values of derivatives, you want to use them; the spline fit might be non-monotonic, so you might want to use a monotonic interpolator like PCHIP or Akima;

All in all I think it'd be better to have a comprehensive tutorial (possibly in the Scipy Cookbook), showing how to subclass `rv_continuous` and how to use it effectively (e.g., use `.rvs(..., size=...)` etc), with several worked examples: what to do if your pdf is sampled, what to do if it's a histogram, how to deal with the tails and so on. I'd be happy to help with the tutorial BTW.

**Comment by rgommers** [comment]
@ev-br the Monte Carlo simulation issue is about using sampled input only right? Do you see value in constructing a distribution instance from a function (this could then be any function that does density estimation)? I kind of like that idea.

**Comment by ev-br** [comment]
@rgommers you mean syntactic sugar like `my_distribution = rv_arbitrary(pdf=my_callable, a=a, b=b)` instead of

```
class my_distribution_gen(rv_continuous):
    def _pdf(self, x, *args):
        return my_callable(x, *args)
my_distribution = my_distribution_gen(a, b)
```

correct? This I won't object to.

**Comment by rgommers** [comment]
Indeed. To make your 4-liner above really nice it needs a bit more though, some of which is in this PR already. Like ``NotImplemented`` for fit methods, getting the docstring right (``rv_continuous.__init__``) defaults are probably not what you want here). I think this needs its own ``__init__`` so the seed can be set, etc.

Or a slightly fancier version could take ``samples`` and use ``gaussian_kde`` or a custom given method on it.

**Comment by ev-br** [comment]
OK, we're on the same page. I still believe a narrative docs with worked examples would be best, but this is off-topic for this PR then :-).

**Comment by andyfaff** [comment]
If I understand the comments correctly there's a consensus about the usefulness of a functional form:

    class my_distribution(rv_continuous):
        def __init__(self, a=-np.inf, b=np.inf, pdf=my_callable,
                          cdf=my_callable2, ppf=my_callable3):
            pass
        def _pdf(self, x, *args):
            return my_callable(x, *args)
        def _cdf(self, x, *args):
            return my_callable2(x, *args)
         def _ppf(self, q, *args):
            return my_callable3(q, *args)

One factor of paramount importance in Monte Carlo simulations is speed. I tested how long it takes to generate 1000 random numbers from the quadratic example in the existing code, it's on the order of 1 second. For most purposes that is way too slow, if you're generating millions or billions of numbers you need to get orders of magnitude faster (I'd say at least 3 or 4). The `_ppf` function needs to be really smart, you don't want to be using `brentq` to examine the CDF for each sample.
Such a generic distribution class could address use cases such as #6419. I can continue on with this PR solely on this use case.

However, what I'm really interested in is generating random numbers from arbitrary distributions described by tabulated`x, y` pairs (possibly with an uncertainty on y), where y is the PDF. In contrast with KDE one creates a PDF from samples; what happens if you don't have the samples? There are a few further issues with KDE:

* one smooths out sharp features in the PDF, e.g. KDE on samples from a uniform distribution have non sharp cutoffs at 0 and 1. There are also problems with oversmoothing on multimodal distributions.
* the KDE PDF can escape the support of the original distribution.

        import numpy as np
        from scipy.stats import gaussian_kde
        r = np.random.random(100000)
        G = gaussian_kde(r)
        g = G.resample(100000)
        # the next line will probably raise an AssertionError
        assert(np.max(g) < 1)
        # this won't
        assert(np.max(r) < 1)

In order to organise the tabulated data I used an `InterpolatedUnivariateSpline`. I had foreseen the use of other types of interpolation, but hadn't worked on them. It's a valid point that one might want to use other interpolators. One possible way around this is to have `interpolator=<subclass UnivariateSpline>, interpolator_args, interpolator_kwds` keywords for tabulated data which says how the interpolation is supposed to be done. The other approach is for the user to create a `UnivariateSpline` object and pass that as a functional form to `my_distribution`.
The reason I went with an `InterpolatedUnivariate` spline is because I could analytically calculate the `_ppf` function where I knew the interpolation was going to be linear (and pass through the data points). If you rely on the generalised`rv_continuous` machinery to calculate PPF it is a factor of x8000 slower - too slow to generate millions of samples. Dealing with those kinds of special cases makes a big difference to usability, one that people in user land may have less ability to implement.

**Comment by josef-pkt** [comment]
http://jpktd.blogspot.ca/2012/12/visual-inspection-of-random-numbers.html

I didn't look much at the details of this PR yet. So just some general comments on the discussion.

IMO, a generic "arbitrary continuous distribution" is too generic to be useful. I think a piecewise linear/constant distribution would cover most of these use cases and could be implemented using relatively fast methods. I wrote a histogram distribution once, with piecewise constant pdf and piecewise linear interpolated cdf and ppf. (I don't think I committed it anywhere.)

Trying to be more generic runs into the problems of how to treat discontinuities (e.g. in expect integration) and the question on how well to approximate an arbitrary function on a continuous support but which need not be continuous, smooth, ...
e.g. generic smooth distributions could be based on series expansion or higher order splines, but it gets much messier

**Comment by ev-br** [comment]
Speaking of the functional form of creating a custom distribution: Andrew's draft sample above does not need to be a subclass of rv_continuous. It can be just a factory function which accepts callables for pdf, ppf, et al, and returns an rv_continuous instance.

**Comment by andyfaff** [comment]
Would this be in a monkey patching fashion?

    def rv_generic(pdf, a, b):
        rv = rv_continuous(a=a, b=b)
        rv._pdf = pdf
        return rv

etc.

**Comment by andyfaff** [comment]
So how about something along these lines, with more documentation. 

* Is it necessary to supply extra `args` to each of `_pdf`, `_cdf` and `_ppf`?
* Would each of those functions need a separate *args?
* Would it be good practice to check if `pdf(x) > 0` for all `x`?
* a previous comment indicates that it should be up to the user to normalise the PDF/CDF, etc.  Is it worth checking this?

Code:

    def func_wrapper(func, args):
        def args_wrapper(x):
            return func(x, *args)
        return args_wrapper

    def rv_generic(pdf, a=None, b=None, ppf=None, cdf=None, args=()):
        # stick in all other kwds for rv_continuous (except shape?)
        rv = rv_continuous(a=a, b=b)
        rv._pdf = func_wrapper(pdf, args)
        if cdf is not None:
            rv._cdf = func_wrapper(cdf, args)
        if ppf is not None:
            rv._ppf = func_wrapper(ppf, args)
        return rv


If one doesn't normalise then one can get funny behaviour, I'm unsure why `rv_continuous.cdf(b)` is being forced to 1:

    def quad(x, *args):
        return 5 * x**2

    a = rv_generic(quad, a=-1, b=1)
    print(a.cdf(0.99999), a.cdf(1))
    # 3.33328333383 1.0

**Comment by ev-br** [comment]
Hmm, I did not think about `*args`, shape parameters. They need to be known at `__init__` time, so you do seem to need a subclass after all. 

cdf(b) is being forced to one in `rv_generic.cdf`

**Comment by andyfaff** [comment]
@ev-br, I don't see why that would require it be a subclass. Can't one just pass in `shapes` and all the other kwds?

    def rv_generic(pdf, a=None, b=None, ppf=None, cdf=None, shapes=None, **kwds):
        rv = rv_continuous(a=a, b=b, shapes=shapes)

**Comment by andyfaff** [comment]
Hmm using the factory function approach would need to monkey patch `rv_generic.freeze` as well.

**Comment by andyfaff** [comment]
I tried using a factory function but I started having to monkey patch a lot of things, and special difficulties were going to be experienced with frozen distributions. I've gone back to subclassing `rv_continuous`. Taking into account the comments above this PR solely adds a distribution defined by a user defined function.
As @josef-pkt mentions above I still think piecewise constant and piecewise linear distributions based on x/y datapoints are important. However, these could easily be done in other code at a later stage (that code would rely on the functionality here).

**Comment by ev-br** [comment]
>I tried using a factory function but I started having to monkey patch a lot of things, and special difficulties were going to be experienced with frozen distributions. I've gone back to subclassing rv_continuous.

Apologies for nudging you towards a wrong path!

**Comment by andyfaff** [comment]
I think it's working reasonably well now, hopefully it just requires polishing. Would it be possible to get someone to have a read over, and let me know if there are major structural changes required.

* One of the major changes is `rv_arbitrary.*` ignores `loc` and `scale`. The rationale being that it's pretty much impossible to keep the pdf normalisation correct if `loc` and `scale` are allowed to vary, for any arbitrary user function. In anycase these can be implemented via `shapes` if required.
* At the moment all the public methods are inheriting from `rv_continuous`. Unfortunately all the methods mention `loc` and `scale` which are ignored. Not sure whether to leave as is, or override all the methods, solely adjusting the docstrings.
* Currently the PDF isn't checked to make sure its normalised. `np.histogram` uses the `normed` kwd. Is this worth adding in (comment above was that it was the users responsibility)?

**Comment by andyfaff** [comment]
Ping @rgommers @ev-br @josef-pkt, any further viewpoints on the future of this?

**Comment by josef-pkt** [comment]
I briefly browsed through the code.

AFAICS, this currently doesn't have any computational shortcuts, like linear interpolation or similar.
So, it doesn't provide any additional functionality compared to just writing a subclass of rv_continuous directly and specifying the `_pdf` and optionally other things.
So, AFAICS, this just provides another API to subclassing, that might make some things automatic or easier. 
Or, do I interpret this in the wrong way?

(My initial impression was that this goes towards interpolation/histogram distribution to make arbitrary code faster. Here, rvs still would go through the looong detour pdf-cdf-ppf, AFAICS)

File: scipy/stats/_continuous_distns.py [hunk 6]
```diff
@@ -5198,8 +5199,280 @@ def _entropy(self, beta):
...
+        shapes = shapes_list[0]
+
+        # make sure the signatures are consistent
+        for item in shapes_list:
+            if item != shapes:
+                raise TypeError('Shape arguments are inconsistent.')
+    else:
+        shapes = []
+
+    return shapes
```
Comment (ev-br) [inline scipy/stats/_continuous_distns.py]: Same here: this is essentially a copy-paste of the logic from `rv_continuous`? If it's really needed here, this helper should be used in both places. Preferably though, none of this should be necessary outside of `rv_continuous.__init__`

File: scipy/stats/tests/test_distributions.py [hunk 7]
```diff
@@ -2274,6 +2275,63 @@ def test_trapz_vect(self):
         assert_allclose(v, res.reshape(v.shape), atol=1e-15)
 
 
+class TestArbitrary(TestCase):
```
Comment (ev-br) [inline scipy/stats/tests/test_distributions.py]: I think a distribution constructed via `rv_arbitrary` should be added to the test loop in `test_continuous.py`.

**Review by ev-br — COMMENTED** [review]
In its current form it's IMO:
1. It should be advertised better. If it's supposed to help with defining new custom distributions, it should be mentioned e.g. at the top of https://github.com/scipy/scipy/blob/master/scipy/stats/__init__.py and in the stats tutorial. As is, it's hiding where it's impossible to find.
2. A large part of implementation is duplicating bits and pieces from the superclass. This is AFAICS mostly to  work around a limitation of `loc=0, scale=1`, and the latter is IMO not justified.
3. An instance of this distribution should be added to the generic loop in https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_continuous_basic.py#L68

**Comment by ev-br** [comment]
Speaking of the piecewise constant pdf, piecewise linear cdf. Here's an example of how that might look within the current framework. Maybe this PR could simplify it?
https://gist.github.com/ev-br/85fe13b6d17fb51a2196da9da7b3ad2f

**Comment by mdhaber** [comment]
Maybe all the new work with fast inversion would help here?

**Pull request closed without merging**